core/trino-main/src/main/java/io/trino/sql/analyzer/CanonicalizationAware.java

```diff
@@ -26,6 +26,11 @@
 {
     private final T node;
 
+    // This class is effectively immutable, even though this field is not final. The hashcode is deterministic, so multiple threads
+    // computing it will produce the same hash. There's a benign data race when updating the hash without synchronization
+    // that is safe because updating an int field is an atomic operation in the JMM.
```

Comment:
I would skip the comment, especially since we're going to have more cases like that.
in particular, probably first such, TypeSignature#hashCode, went without such comment.

if you want a comment, make it a short one liner, so it doesn't feel too heavy to copy to every other such place (see @ppalucha 's https://github.com/trinodb/trino/pull/9345)

Reply:
btw worth adding a node in io.trino.sql.tree.Node that impl must be immutable

Reply:
I think a comment is warranted -- it's not immediately obvious that the class is properly thread-safe without any form of synchronizations.

For #9345, I'm still not convinced we should add such logic to every single node in the AST. An approach list ScopeAware or CanonicalizationAware is more future-proof / fool-proof.

Reply:
There's a nice explanation of what is required in String class source - https://github.com/AdoptOpenJDK/openjdk-jdk16/blob/master/src/java.base/share/classes/java/lang/String.java#L1531.

Reply:
> I think a comment is warranted -- it's not immediately obvious that the class is properly thread-safe without any form of synchronization.

@martint 
i don't think it's needed, but sure, you can have it. As i said, please make it short, ideally one-liner. Just a **hint** that it's concurrently safe is enough.
```
// This field is mutable, but class remains effectively immutable
```


@ppalucha thanks for the link.
i am surprised they have (added?) `hashIsZero` field. 
Having two fields requires longer explanation.

Reply:
> Just a hint that it's concurrently safe is enough.

This falls in the realm of obscure JMM semantics and tricky concurrent code. I don't think glossing over it is a good idea. It may lead to people cargo-cult copying what this is doing in other places without understanding why. Also, we should keep this kind of pattern confined to just a few places, not spread out throughout the codebase (related to my comment in the other PR regarding how we should cache hashcodes for expressions)

Reply:
> This falls in the realm of obscure JMM semantics and tricky concurrent code.

I don't think atomic write to an `int` field is tricky or obscure.

Reply:
It's more than atomic write to an int. The correctness of that code depends on:
* The computation is idempotent and derived from safely-publicated data (final fields)
* The value read/written to that field is observed atomically
* The function reads from the field once, and stores the result once
* We don't care if multiple threads (or every thread, for that matter) computes the hash. Per the JMM, that's a valid outcome due to lack of synchronization

It is obscure because it's relying on a data race, for which the JMM has fewer guarantees. In case you haven't seen it, this describes all sorts of gotchas you might run into when you cheat or deviate from the rules: https://shipilev.net/blog/2016/close-encounters-of-jmm-kind/